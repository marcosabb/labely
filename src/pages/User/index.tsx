import React, { useState, useCallback } from 'react'
import { useNavigation } from '@react-navigation/native'

import { useUsers } from '../../contexts/users'

import Input from '../../components/Input'
import Button from '../../components/Button'

import {
  Container,
  Header,
  GithubIcon,
  Form,
  Title,
  Description,
  Field,
  Policies,
  PoliciesText
} from './styles'

export default function Home() {
  const navigation = useNavigation()
  const { createUser, loading } = useUsers()

  const [user, setUser] = useState('')

  const handleChange = useCallback((value: string) => {
    setUser(value)
  }, [])

  const handleSubmit = useCallback(async () => {
    if (!user) return

    await createUser(user)
    navigation.navigate('Users')
  }, [createUser, navigation, user])

  return (
    <Container>
      <Header>
        <GithubIcon />
      </Header>

      <Form>
        <Title>Buscar usuário</Title>
        <Description>
          Crie sua conta através do seu usuário do GitHub
        </Description>

        <Field>
          <Input
            name='user'
            placeholder='Digite o usuário'
            icon='account-circle'
            onChangeText={handleChange}
            testID='user-input'
          />
        </Field>

        <Button
          onPress={handleSubmit}
          loading={loading.actions}
          disabled={!user}
          testID='user-button'
        >
          Cadastrar
        </Button>
      </Form>

      <Policies>
        <PoliciesText>Termos de política e privacidade</PoliciesText>
      </Policies>
    </Container>
  )
}
