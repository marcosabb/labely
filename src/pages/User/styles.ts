import styled from 'styled-components/native'
import { theme } from 'styled-tools'

export const Container = styled.View`
  flex: 1;
  justify-content: space-between;
  padding: 0 ${theme('spacing.xlg')} ${theme('spacing.lg')};
  background-color: ${theme('colors.foreground')};
`

export const Header = styled.View`
  align-items: center;
`

export const Form = styled.View``

export const Title = styled.Text`
  margin-bottom: ${theme('spacing.sm')};
  font-family: ${theme('typography.families.bold')};
  font-size: ${theme('typography.sizes.md')};
  color: ${theme('colors.title')};
`

export const Description = styled.Text`
  margin-bottom: ${theme('spacing.md')};
  font-family: ${theme('typography.families.regular')};
  font-size: ${theme('typography.sizes.default')};
  color: ${theme('colors.text')};
`

export const Field = styled.View`
  margin-bottom: ${theme('spacing.md')};
`

export const Policies = styled.TouchableOpacity`
  align-items: center;
  justify-content: center;
`

export const PoliciesText = styled.Text`
  font-family: ${theme('typography.families.regular')};
  font-size: ${theme('typography.sizes.default')};
  color: ${theme('colors.text')};
`
