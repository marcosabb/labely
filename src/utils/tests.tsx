import React from 'react'
import { render, RenderOptions } from '@testing-library/react-native'
import { ThemeProvider } from 'styled-components/native'

import theme from 'styles/theme'

interface Props {
  children: React.ReactNode
}

function ThemeWrapper({ children }: Props) {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: ThemeWrapper, ...options })

export * from '@testing-library/react-native'

export { customRender as render }
