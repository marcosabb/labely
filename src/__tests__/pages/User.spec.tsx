import React from 'react'
import { fireEvent, waitFor } from '@testing-library/react-native'

import User from 'pages/User'
import { render } from 'utils/tests'

const mockedNavigate = jest.fn()

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: mockedNavigate,
    setOptions: () => ({
      headerRight: () => {}
    })
  }),
  useRoute: () => ({
    params: {}
  })
}))

jest.mock('../../contexts/users', () => ({
  useUsers: () => ({
    users: [
      {
        id: 56012991,
        name: 'Marcos Borges',
        avatar_url: 'https://avatars.githubusercontent.com/u/56012991?v=4',
        login: 'marcosabb',
        company: 'Incentivar.io',
        location: 'BrasÃ­lia, Brazil'
      }
    ],
    loading: {},
    createUser: jest.fn()
  })
}))

describe('Users page', () => {
  beforeEach(() => {
    mockedNavigate.mockClear()
  })

  it('should match snapshot', () => {
    const component = render(<User />)

    expect(component).toMatchSnapshot()
  })

  it('should be able to add new user', async () => {
    const { getByTestId } = render(<User />)

    const input = getByTestId('user-input')
    const button = getByTestId('user-button')

    fireEvent.changeText(input, 'marcosabb')

    fireEvent.press(button)

    await waitFor(() => {
      expect(mockedNavigate).toHaveBeenCalledWith('Users')
    })
  })

  it('should not be able to add new user if name was not provided', async () => {
    const { getByTestId } = render(<User />)

    const input = getByTestId('user-input')
    const button = getByTestId('user-button')

    fireEvent.changeText(input, '')

    fireEvent.press(button)

    await waitFor(() => {
      expect(mockedNavigate).not.toHaveBeenCalledWith('Users')
    })
  })
})
