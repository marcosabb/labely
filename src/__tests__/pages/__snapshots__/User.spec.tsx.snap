// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users page should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "space-between",
        "paddingBottom": 32,
        "paddingLeft": 40,
        "paddingRight": 40,
        "paddingTop": 0,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 100,
          },
          Array [
            Object {},
          ],
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󰊤
    </Text>
  </View>
  <View
    style={
      Array [
        Object {},
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "Mulish-Bold",
            "fontSize": 22,
            "marginBottom": 10,
          },
        ]
      }
    >
      Buscar usuário
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#7e7e7e",
            "fontFamily": "Mulish-Regular",
            "fontSize": 16,
            "marginBottom": 24,
          },
        ]
      }
    >
      Crie sua conta através do seu usuário do GitHub
    </Text>
    <View
      style={
        Array [
          Object {
            "marginBottom": 24,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderColor": "#e5e5e5",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 40,
              "paddingLeft": 16,
              "paddingRight": 16,
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#e5e5e5",
                "fontSize": 24,
              },
              Array [
                Object {},
              ],
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          name="user"
          onChangeText={[Function]}
          placeholder="Digite o usuário"
          placeholderTextColor="#7e7e7e"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "color": "#000",
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "fontFamily": "Mulish-Regular",
                "fontSize": 16,
                "height": 40,
                "marginLeft": 8,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              },
            ]
          }
          testID="user-input"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#000",
          "borderRadius": 100,
          "height": 40,
          "justifyContent": "center",
          "opacity": 0.8,
        }
      }
      testID="user-button"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontFamily": "Mulish-Regular",
              "fontSize": 16,
            },
          ]
        }
      >
        Cadastrar
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#7e7e7e",
            "fontFamily": "Mulish-Regular",
            "fontSize": 16,
          },
        ]
      }
    >
      Termos de política e privacidade
    </Text>
  </View>
</View>
`;
